Databases tree view

danvergara opened the pull request:
# Add a view to see databases and their tables

## Description

The vast majority of database client software, if not all, show a list of databases and their tables as a tree structure. This is convenient for users that want to see all the database the have access to.
When it comes to dblab, things used to be a bit different, because database parameter was always required, either as a cli flag value or as part of the DSN. This was this way for the longest time because one reason. First off, `dblab` was heavily inspired by [pgweb](https://sosedoff.github.io/pgweb/). This was the client I used to use at work, but suddenly I got tasked to manage a couple of microservices tied to MySQL databases, and I needed a FOSS alternative for Linux that worked as well as `pgweb` for my use case. `pgweb` rquires a database to connect to and only display its tables as a list. Then I tried to mimic a pgweb like client in the terminal and that's how dblab started. But things have changed I my needs have evolved and sometimes I need to look at other databases as well. 

This MR basically removes the mandatory status of the database parameter and makes it optional. If the database is not provided, then, a tree structure is displayed  showig databases and their tables.

Another important change has been made and it was removing the pagination. I have rarely found myself using it, because it's faster to make query to where I need to be.

![Screenshot from 2024-11-15 23-36-45](https://github.com/user-attachments/assets/f411113d-9eab-4387-a7fd-7f15a770ca58)

![Screenshot from 2024-11-15 23-42-24](https://github.com/user-attachments/assets/3a64be73-8501-4a38-ad18-22496ce35fab)


Fixes #209 
Fixes #232

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Tested it against pagila databse and the testing database of the project

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

ccoVeille commented on pkg/client/client.go:
```diff
@@ -93,15 +119,49 @@ func New(opts command.Options) (*Client, error) {
 // Query returns performs the query and returns the result set and the column names.
 func (c *Client) Query(q string, args ...interface{}) ([][]string, []string, error) {
-	resultSet := [][]string{}
+	var (
+		resultSet = [][]string{}
+		db        *sqlx.DB
+	)
+
+	if c.activeDatabase != "" {
+		switch c.driver {
+		case drivers.Postgres, drivers.PostgreSQL, drivers.MySQL:
+			db = c.dbs[c.activeDatabase]
+		}
+	} else {
+		db = c.db
+	}
```
What about this?


```suggestion
	var (
		resultSet = [][]string{}
	)

	db = c.db
	if c.activeDatabase != "" {
		switch c.driver {
		case drivers.Postgres, drivers.PostgreSQL, drivers.MySQL:
			db = c.dbs[c.activeDatabase]
		}
	}
```

danvergara replied:
This is so nice and elegant. Good catch!

ccoVeille commented on pkg/client/client.go:
```diff
@@ -93,15 +119,49 @@ func New(opts command.Options) (*Client, error) {
+	if c.activeDatabase != "" {
+		switch c.driver {
+		case drivers.Postgres, drivers.PostgreSQL, drivers.MySQL:
+			db = c.dbs[c.activeDatabase]
```
What about this?


```suggestion
			db, ok = c.dbs[c.activeDatabase]
			if !ok {
	 			return nil, nil, return nil, fmt.Errorf("connection with %s database not found", c.activeDatabase)
			}
```

I'm afraid about possible panic, when calling db.Query

danvergara replied:
Yeah, good catch there!

I'll add this.

Thanks!

ccoVeille commented on pkg/client/client.go:
```diff
@@ -257,69 +305,68 @@ func (c *Client) ShowTables() ([]string, error) {
-	if err := c.paginationManager.PreviousPage(); err != nil {
-		return nil, 0, err
+		tables = append(tables, table)
 	}
```
Each row.Nect loop must be followed by checking rows.Err()

```suggestion
	}

 	if err:= rows.Err(); err != nil {
 	 	return nil, err
 	}
```

ccoVeille commented on pkg/client/client.go:
```diff
@@ -257,69 +305,68 @@ func (c *Client) ShowTables() ([]string, error) {
-	return c.db
-}
+		databases = append(databases, database)
+	}
```
Same issue about missing rows.Err check

ccoVeille commented on pkg/client/client.go:
```diff
@@ -416,3 +442,27 @@ func (c *Client) indexes(tableName string) ([][]string, []string, error) {
+func getDB(driver, connString, database string) (*sqlx.DB, error) {
+	var newConnString string
+
+	if driver == drivers.MySQL {
+		newConnString = strings.Replace(connString, "/", fmt.Sprintf("/%s", database), 1)
+	} else {
+		u, err := url.Parse(connString)
+		if err != nil {
+			return nil, err
+		}
+
+		u.Path = "/" + database
+		newConnString = u.String()
+	}
```
I would use a switch and return an error when it's none of the supported database types

I'm a bit worried by the fact you will apply the postgres way to any future database type that might be added

danvergara replied:
I agree with using a switch here, but the driver validation is done somewhere. If the execution reaches this point, the driver is actually valid. Regarding applying the postgres way, yeah a switch statement might help supporting a new use case.

ccoVeille commented on pkg/client/client.go:
```diff
@@ -416,3 +442,27 @@ func (c *Client) indexes(tableName string) ([][]string, []string, error) {
+	db, err := sqlx.Open(driver, newConnString)
+	if err != nil {
+
+		panic(err)
```
Why a panic?


```suggestion
		return nil, err
```

danvergara replied:
I left it there by accident. I'll remove it,
Thanks!

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -39,7 +39,15 @@ func TestMain(m *testing.M) {
 func generateURL(driver string) string {
 	switch driver {
 	case drivers.Postgres:
```
Why postgresql const is not present here? Isn't it tested?

danvergara replied:
Because the that constant is the one used at the time to test on CI. I can only use one at a time there.

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -276,14 +279,17 @@ func TestIndexes(t *testing.T) {
+	t.Logf("constraints content %v", r)
 
 	assert.NoError(t, err)
 	assert.Greater(t, len(r), 0)
 	assert.Greater(t, len(co), 0)
```
```suggestion
 	assert.NotEmpty(t, r)
 	assert.NotEmpty(t, co)
```

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -302,10 +308,11 @@ func TestCountTable(t *testing.T) {
 
-	assert.Equal(t, count, 100)
 	assert.NoError(t, err)
+	assert.Greater(t, len(r), 0)
+	assert.Greater(t, len(co), 0)
```
Same you could use NotEmpty

ccoVeille commented on pkg/tui/tui.go:
```diff
@@ -351,17 +413,17 @@ func (t *Tui) updateTableMetadataOnChange(tableName string) {
-			} else {
-				t.aw.content.SetCellSimple(i+1, j, sc)
-			}
+			// if i == 0 {
+			// 	t.aw.content.SetCell(i+1, j, &tview.TableCell{Text: sc, Color: tcell.ColorRed})
+			// } else {
+			// }
+			t.aw.content.SetCellSimple(i+1, j, sc)
```
Why leaving commented out code?

danvergara replied:
Good catch!

I was testing out new colors.

danvergara commented:
Hey @ccoVeille!

What a good code review there!

I already addressed all your comments. Let me know if I missed something.

Thanks so much!

P.S. Can I include you as a reviewer from this point on?

ccoVeille commented:
> Hey @ccoVeille!
> 
> What a good code review there!

Thanks, you are welcome
 
> I already addressed all your comments. Let me know if I missed something.
> 
> Thanks so much!
> 
> P.S. Can I include you as a reviewer from this point on?

Sure

ccoVeille commented:
Maybe you could add a [co-authored-by](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) tag

You can find my email here

https://github.com/samber/lo/commit/52a73c3ffda944e42c2d6f4fc1fb06a8eaf114c6

So, just add this to the commit message of the changes you made with my review

```
Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>
```

ccoVeille commented on pkg/client/client.go:
```diff
@@ -416,3 +461,27 @@ func (c *Client) indexes(tableName string) ([][]string, []string, error) {
+	switch driver {
+	case drivers.MySQL:
+		newConnString = strings.Replace(connString, "/", fmt.Sprintf("/%s", database), 1)
+	default:
+		u, err := url.Parse(connString)
+		if err != nil {
+			return nil, err
+		}
+
+		u.Path = "/" + database
+		newConnString = u.String()
```
This would be safer (I might have forgotten some)

```suggestion
	case drivers.PostgreSQL, drivers.Postgres:
		u, err := url.Parse(connString)
		if err != nil {
			return nil, err
		}

		u.Path = "/" + database
		newConnString = u.String()

	default:
		return nil, fmt.Errorf("unsupported driver %s", driver)
```

danvergara replied:
I disagree with this, since the driver validation is done somewhere else and the MySQL URL parsing is the corner case here The default behavior is parsing the `DSN` with `url.Parse`.

ccoVeille replied:
Ah OK then. I misunderstood

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -332,9 +339,6 @@ func TestMetadata(t *testing.T) {
-	assert.Equal(t, m.TotalPages, 1)
-
 	// indexes.
 	assert.Greater(t, len(m.Indexes.Rows), 0)
 	assert.Greater(t, len(m.Indexes.Columns), 0)
```
Use NotEmpty

ccoVeille replied:
testifylint linter could help you a lot with such "issue"

It would also detect when you should use require.NoError

https://github.com/Antonboom/testifylint

It can be added easily to golangci-lint, bit you don't have golangci-lint yet,
maybe

I can help you here if you want

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -193,15 +201,16 @@ func TestShowTables(t *testing.T) {
 	}
 
-	c, _ := New(opts)
+	c, err := New(opts)
+	assert.NoError(t, err)
 
 	tables, err := c.ShowTables()
```
here you should (must) use

```suggestion
	c, err := New(opts)
	require.NoError(t, err)

	tables, err := c.ShowTables()
```

require is another package available in testify

The reason is that if an error is returned by New, c will be nil, so c.ShowTables will panic

Test should never panic

danvergara replied:
Ok, let's focus on the scope of the PR. We can address the test suite in another PR and issue. This test suite is not new, it's been 3 years since I set it up. I'm willing to improve it but latter.

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -212,23 +221,20 @@ func TestTableStructure(t *testing.T) {
 		Limit:  100,
 	}
 
 	c, _ := New(opts)
```
You shouldn't ignore the error here

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -247,17 +253,14 @@ func TestConstraints(t *testing.T) {
 
 	c, _ := New(opts)
```
Same

ccoVeille commented on pkg/client/client_test.go:
```diff
@@ -276,14 +279,17 @@ func TestIndexes(t *testing.T) {
-	r, co, err := c.indexes("products")
+	r, co, err := c.constraints("products")
+
+	t.Logf("constraints columns %v", co)
+	t.Logf("constraints content %v", r)
```
These logs are not needed. If the test fail, you will get the info. If it doesn't fail, I don't see a need to see these lines when running a `go test`

danvergara replied:
Let's create an issue to improve the test suite.

danvergara commented:
> Maybe you could add a [co-authored-by](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) tag
> 
> You can find my email here
> 
> [samber/lo@52a73c3](https://github.com/samber/lo/commit/52a73c3ffda944e42c2d6f4fc1fb06a8eaf114c6)
> 
> So, just add this to the commit message of the changes you made with my review
> 
> ```
> Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>
> ```

Alright @ccoVeille, I just added the co-authored tag. Honor when honor is due.

Can you go over the PR again, please? I 'd like to land this change tonight.

Arturomtz8 reviewed (approved):
nice work man
